### PR TITLE
Feature/pbac support for cases

### PR DESCRIPTION
**Pull request by ivo-ritense**

### CODE REVIEW

File: app/gzac/src/main/java/com/ritense/gzac/listener/ApplicationReadyEventListener.kt [hunk 1]
```diff
@@ -395,40 +398,50 @@ class ApplicationReadyEventListener(
...
                     roleKey = USER
                 ),
                 Permission(
                     resourceType = JsonSchemaDocument::class.java,
-                    action = Action.CLAIM,
+                    action = JsonSchemaDocumentActionProvider.CLAIM,
                     conditionContainer = ConditionContainer(
                         listOf(
-                            FieldPermissionCondition("assigneeId", EQUAL_TO, "\${currentUserId}")
+                            FieldPermissionCondition("assigneeFullName", EQUAL_TO, "James Vance")
```
Comment (Klaas-Ritense) [inline app/gzac/src/main/java/com/ritense/gzac/listener/ApplicationReadyEventListener.kt:414]: please change

File: document/src/test/resources/config/application.yml [hunk 2]
```diff
@@ -8,8 +8,8 @@ spring:
         hikari:
             auto-commit: false
     liquibase:
-        enabled: true
-        change-log: classpath:/config/liquibase/document-master.xml
+        enabled: false
+#        change-log: classpath:/config/liquibase/document-master.xml
```
Comment (Klaas-Ritense) [inline document/src/test/resources/config/application.yml:12]: Won't this break the tests?

File: form/src/main/kotlin/com/ritense/form/service/FormSubmissionService.kt [hunk 3]
```diff
@@ -77,26 +78,29 @@ open class FormSubmissionService(
...
-                processLink,
-                document,
-                taskInstanceId,
-                processDocumentDefinition,
-                submittedDocumentContent,
-                formDefinedProcessVariables,
-                preJsonPatch
-            )
-            return dispatchRequest(request, processDocumentDefinition, formFields, externalFormData)
+            return AuthorizationContext.runWithoutAuthorization {
```
Comment (Klaas-Ritense) [inline form/src/main/kotlin/com/ritense/form/service/FormSubmissionService.kt:81]: Why is authorization disabled here? not everyone is allowed to submit a form

File: document/src/test/resources/config/application.yml (hunk 2 above)
Comment (marijnritense) [reply document/src/test/resources/config/application.yml:12]: Nope, tests have been fixed. The problem with how it was set up here is none of the liquibase changesets from other modules would run out of the box. This is fine if all we're testing is just document itself, but we aren't when including authorization calls.

File: form/src/main/kotlin/com/ritense/form/service/FormSubmissionService.kt (hunk 3 above)
Comment (marijnritense) [reply form/src/main/kotlin/com/ritense/form/service/FormSubmissionService.kt:81]: Endpoint itself is still protected, just like it was before. Support for forms will not be included in major 11.

File: document/src/main/java/com/ritense/document/service/impl/JsonSchemaDocumentSearchService.java [hunk 4]
```diff
@@ -165,40 +154,45 @@ private Page<JsonSchemaDocument> search(QueryWhereBuilder queryWhereBuilder, Pag
...
+        queryWhereBuilder.apply(cb, countQuery, countRoot);
 
+        // TODO: Should be turned into a subquery, and then do a count over the results from the subquery.
         List<Long> countResultList = entityManager.createQuery(countQuery).getResultList();
 
         Long count = 0L;
 
         if (!countResultList.isEmpty()) {
-            count = countResultList.get(0);
+            count = (long) countResultList.size();
```
Comment (theo-ritense) [inline document/src/main/java/com/ritense/document/service/impl/JsonSchemaDocumentSearchService.java:175]: Wont this always result in `1`? There's usually only one result in a count query.

File: document/src/main/kotlin/com/ritense/document/service/JsonSchemaDocumentSpecification.kt [hunk 5]
```diff
@@ -57,4 +61,16 @@ class JsonSchemaDocumentSpecification(
...
         return combinePredicates(criteriaBuilder, predicates)
     }
+
+    companion object {
+        @JvmStatic
+        fun byDocumentDefinitionIdName(name: String): Specification<JsonSchemaDocument> {
+            return Specification { root: Root<JsonSchemaDocument>,
+                                   _: CriteriaQuery<*>?,
+                                   criteriaBuilder: CriteriaBuilder ->
+                // documentRoot.get(DOCUMENT_DEFINITION_ID).get(NAME), documentDefinitionName)
```
Comment (theo-ritense) [inline document/src/main/kotlin/com/ritense/document/service/JsonSchemaDocumentSpecification.kt:71]: Please remove the commented code. We have git for that.

File: document/src/test/java/com/ritense/document/web/rest/searchfields/SearchFieldResourceIntegrationTest.java [hunk 6]
```diff
@@ -179,6 +187,7 @@ void shouldRespondWith400BadRequestOnSearchFieldPathValidationError() throws Exc
     }
 
     @Test
+    @WithMockUser(username = USERNAME, authorities = FULL_ACCESS_ROLE)
```
Comment (theo-ritense) [inline document/src/test/java/com/ritense/document/web/rest/searchfields/SearchFieldResourceIntegrationTest.java:190]: The others use `{FULL_ACCESS_ROLE}`?

File: notes/src/test/kotlin/com/ritense/note/web/rest/NoteResourceIT.kt [hunk 7]
```diff
@@ -131,29 +134,29 @@ internal class NoteResourceIT : BaseIntegrationTest() {
...
-                .contentType(APPLICATION_JSON_VALUE)
-                .content(jacksonObjectMapper().writeValueAsString(note))
-        )
-            .andDo(print())
-            .andExpect(status().isForbidden)
-    }
+    // TODO: implement authorization for note
+//    @Test
+//    @WithMockUser(TEST_USER)
+//    fun `should not create note when user has no permission to the document`() {
```
Comment (theo-ritense) [inline notes/src/test/kotlin/com/ritense/note/web/rest/NoteResourceIT.kt:140]: That's one way to fix tests.

File: process-document/src/main/java/com/ritense/processdocument/service/impl/CamundaProcessJsonSchemaDocumentAssociationService.java [hunk 8]
```diff
@@ -222,6 +232,7 @@ public void deleteProcessDocumentDefinition(String documentDefinitionName) {
 
     @Transactional
     @Override
+    //TODO: ??? @Marijn
```
Comment (theo-ritense) [inline process-document/src/main/java/com/ritense/processdocument/service/impl/CamundaProcessJsonSchemaDocumentAssociationService.java:235]: Very helpful.

File: document/src/main/java/com/ritense/document/service/impl/JsonSchemaDocumentSearchService.java (hunk 4 above)
Comment (marijnritense) [reply document/src/main/java/com/ritense/document/service/impl/JsonSchemaDocumentSearchService.java:175]: Unless there's a group by, unfortunately.

File: document/src/test/java/com/ritense/document/web/rest/searchfields/SearchFieldResourceIntegrationTest.java (hunk 6 above)
Comment (marijnritense) [reply document/src/test/java/com/ritense/document/web/rest/searchfields/SearchFieldResourceIntegrationTest.java:190]: Varies. Either notation is fine I think.

File: app/gzac/src/main/java/com/ritense/gzac/listener/ApplicationReadyEventListener.kt (hunk 1 above)
Comment (marijnritense) [reply app/gzac/src/main/java/com/ritense/gzac/listener/ApplicationReadyEventListener.kt]: done

File: app/gzac/src/main/java/com/ritense/gzac/listener/ApplicationReadyEventListener.kt [hunk 9]
```diff
@@ -395,40 +398,50 @@ class ApplicationReadyEventListener(
                 ),
                 Permission(
                     resourceType = JsonSchemaDocument::class.java,
-                    action = Action.VIEW,
+                    action = JsonSchemaDocumentActionProvider.VIEW,
                     conditionContainer = ConditionContainer(
                         listOf(
-                            FieldPermissionCondition("assigneeId", EQUAL_TO, "\${currentUserId}")
+                            FieldPermissionCondition("assigneeFullName", EQUAL_TO, "James Vance")
```
[the remark this answers is not in the document]
Comment (marijnritense) [reply app/gzac/src/main/java/com/ritense/gzac/listener/ApplicationReadyEventListener.kt]: done

File: process-document/src/main/java/com/ritense/processdocument/service/impl/CamundaProcessJsonSchemaDocumentAssociationService.java (hunk 8 above)
Comment (marijnritense) [reply process-document/src/main/java/com/ritense/processdocument/service/impl/CamundaProcessJsonSchemaDocumentAssociationService.java]: Changed the comments to be a bit more informative.

File: document/src/main/kotlin/com/ritense/document/service/JsonSchemaDocumentSpecification.kt (hunk 5 above)
Comment (marijnritense) [reply document/src/main/kotlin/com/ritense/document/service/JsonSchemaDocumentSpecification.kt]: removed it

**Review by theo-ritense — APPROVED** [review]

**Pull request merged**